### DISCUSSION TITLE
Lifecycle hook upon asyncReplace update or completion

### DISCUSSION THREAD

**Discussion by laurentpellegrino** [lit/lit#2219] [Q&A] [answered]
I am trying to register a custom click listener for all links that are part of a DOM subtree appended in the lit `render` method using the directive `asyncReplace` after some asynchronous transformations.

The problem I have is that the `updated` lifecycle method is not called upon `asyncReplace` updates or completion. 

Here is a [minimal live example](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudCwgUHJvcGVydHlWYWx1ZXMsIFRlbXBsYXRlUmVzdWx0fSBmcm9tICdsaXQnO1xuaW1wb3J0IHthc3luY1JlcGxhY2V9IGZyb20gJ2xpdC9kaXJlY3RpdmVzL2FzeW5jLXJlcGxhY2UuanMnO1xuaW1wb3J0IHt1bnNhZmVIVE1MfSBmcm9tICdsaXQvZGlyZWN0aXZlcy91bnNhZmUtaHRtbC5qcyc7XG5pbXBvcnQge2N1c3RvbUVsZW1lbnQsIHByb3BlcnR5fSBmcm9tICdsaXQvZGVjb3JhdG9ycy5qcyc7XG5pbXBvcnQge3N0YXRlfSBmcm9tICdsaXQvZGVjb3JhdG9ycy5qcyc7XG5cblxuQGN1c3RvbUVsZW1lbnQoJ3NpbXBsZS1ncmVldGluZycpXG5leHBvcnQgY2xhc3MgU2ltcGxlR3JlZXRpbmcgZXh0ZW5kcyBMaXRFbGVtZW50IHtcblxuICBAc3RhdGUoKVxuICBwcm90ZWN0ZWQgcHVibGljYXRpb25EYXRhOiBBcnJheTxhbnk-ID0gW107XG4gIFxuICBjb25zdHJ1Y3RvcigpIHtcbiAgICBzdXBlcigpO1xuICAgIHRoaXMucHVibGljYXRpb25EYXRhID0gW1xuICAgICAgaHRtbGAke3Vuc2FmZUhUTUwoXCI8ZGl2IGNsYXNzPSdwdWJsaWNhdGlvbic-QTwvZGl2PlwiKX1gLCBcbiAgICAgIGh0bWxgJHt1bnNhZmVIVE1MKFwiPGRpdiBjbGFzcz0ncHVibGljYXRpb24nPkI8L2Rpdj5cIil9YFxuICAgIF07XG4gIH1cbiAgXG4gIHJlbmRlcigpIHtcbiAgICByZXR1cm4gaHRtbGA8ZGl2PllvdXIgcHVibGljYXRpb25zPC9kaXY-JHthc3luY1JlcGxhY2UodGhpcy50cmFuc2Zvcm1QdWJsaWNhdGlvbnMoKSl9YDtcbiAgfVxuICBcbiAgdXBkYXRlZChjaGFuZ2VkUHJvcGVydGllczogUHJvcGVydHlWYWx1ZXMpIHtcbiAgICBzdXBlci51cGRhdGVkKGNoYW5nZWRQcm9wZXJ0aWVzKTtcbiAgICBjb25zb2xlLmxvZygnc2hhZG93Um9vdCcsIHRoaXMuc2hhZG93Um9vdC5pbm5lckhUTUwpO1xuICAgIGNvbnNvbGUubG9nKCdwdWJsaWNhdGlvbnMnLCB0aGlzLnNoYWRvd1Jvb3QucXVlcnlTZWxlY3RvckFsbCgnLnB1YmxpY2F0aW9uJykpO1xuICB9XG4gIFxuICBwcml2YXRlIGFzeW5jKiB0cmFuc2Zvcm1QdWJsaWNhdGlvbnMoKTogQXN5bmNHZW5lcmF0b3I8VGVtcGxhdGVSZXN1bHRbXSwgdm9pZCwgdW5rbm93bj4ge1xuICAgIGF3YWl0IGRlbGF5KDEwMDApO1xuICAgIHlpZWxkIHRoaXMucHVibGljYXRpb25EYXRhLm1hcChlID0-IGUpO1xuICAgIFxuICB9XG59XG5cbmZ1bmN0aW9uIGRlbGF5KG1zOiBudW1iZXIpIHtcbiAgICByZXR1cm4gbmV3IFByb21pc2UoIHJlc29sdmUgPT4gc2V0VGltZW91dChyZXNvbHZlLCBtcykgKTtcbn0ifSx7Im5hbWUiOiJpbmRleC5odG1sIiwiY29udGVudCI6IjwhRE9DVFlQRSBodG1sPlxuPGhlYWQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vc2ltcGxlLWdyZWV0aW5nLmpzXCI-PC9zY3JpcHQ-XG48L2hlYWQ-XG48Ym9keT5cbiAgPHNpbXBsZS1ncmVldGluZyBuYW1lPVwiV29ybGRcIj48L3NpbXBsZS1ncmVldGluZz5cbjwvYm9keT5cbiJ9LHsibmFtZSI6InBhY2thZ2UuanNvbiIsImNvbnRlbnQiOiJ7XG4gIFwiZGVwZW5kZW5jaWVzXCI6IHtcbiAgICBcImxpdFwiOiBcIl4yLjAuMFwiLFxuICAgIFwiQGxpdC9yZWFjdGl2ZS1lbGVtZW50XCI6IFwiXjEuMC4wXCIsXG4gICAgXCJsaXQtZWxlbWVudFwiOiBcIl4zLjAuMFwiLFxuICAgIFwibGl0LWh0bWxcIjogXCJeMi4wLjBcIlxuICB9XG59IiwiaGlkZGVuIjp0cnVlfV0) that logs when `updated` is called. What we can see is that updated is called before `asyncReplace` appends the nodes to DOM but never after. As a result, I cannot retrieve the DOM subtree I would like with a query selector.

Is there a way to have a callback called upon `asyncReplace` update or completion?

**Comment by justinfagnani** [comment, marked as answer]
This is one reason why reactive controllers are in some ways a more powerful tool to handle async values than `until` and `asyncReplate`. The downside is that controllers trigger entire component updates (which is the point here - so you can tie into the update lifecycle) whereas directives update only one binding in a template. You could make a reactive controller to use instead of asyncReplace.

This is untested, but roughly the idea:

```ts
class AsyncValue implements ReactiveController {
  private _host: ReactiveControllerHost;  
  private _iterable?: AsyncIterable<T>;
  private _cancelIteration?: () => void;
  
  value?: T;
  
  set iterable(v?: AsyncIterable<T>) {
    if (this._iterable !== v) {
      this._iterable = v;
      this._iterate();
    }
  }
  
  constructor(host: ReactiveControllerHost) {
    this._host = host;
    host.addController(this);
  }
  
  private async _iterate() {
    const iterable = this._iterable;
    this._cancelIteration?.();
    const cancelPromise = new Promise((res, rej) => this._cancelIteration = rej; );
    for await (const v of Promise.race([iterable, cancelPromise])) {
      if (iterable !== this._iterable) {
        return;
      }
      this.value = v;
      this._host.requestUpdate();
    }
  }
}
```

Then:

```ts
class MyElement extends LitElement {
  private _myValue = new AsyncValue(this);
  
  @property({attribute: false})
  someIterable;
  
  render() {
    this._myValue.iterable = this.someIterable;
    return html`<p>${this._myValue.value}</p>`;
  }
}
```

This is a bit more verbose than a directive, so you could also wrap this up in an async directive to create the controller automatically.

Another option would be to use something like https://github.com/ReactiveX/IxJS to wrap the iterable with a callback that calls `this.requestUpdate()`.

**Reply by laurentpellegrino** [reply]
Thanks!
